### PR TITLE
Support for any message type

### PR DESCRIPTION
**Pull request by Aniket21mathur**
SRC -> https://github.com/Aniket21mathur/ProtocolBuffers-Examples/blob/master/chapel/src/anytype.proto
OUT -> https://github.com/Aniket21mathur/ProtocolBuffers-Examples/blob/master/chapel/build/any.chpl

### CODE REVIEW

**Comment by Aniket21mathur** [comment]
@mppf @lydia-duncan @astatide -> Base implementation and tests for any. Though a couple of small things are left which include:

- Throwing an error if the type specified in URL and the type of the message object passed by the user do not match while deserializing.
- Handling the case when a package name is specified in the .proto file since the general format of the URL is `type.googleapis.com/_packagename_._messagename_`. 

Regarding the second point, I guess there will be a need to have a `string` field `packageName`, or something like that, in our generated record so that we can have the package name of a message object attached to it.  What do you guys think?

**Comment by mppf** [comment]
> Handling the case when a package name is specified in the .proto file since the general format of the URL is type.googleapis.com/_packagename_._messagename_.

You could add a `param typeURL: string = "something-generated";` field to the generated record for all messages and use it to check these types when using any.

**Comment by Aniket21mathur** [comment]
> You could add a `param typeURL: string = "something-generated";`

Instead of `typeURL` I would like to use something more generic for param name, `typeURL` is specific to `Any` I guess. 

Also, since `packFrom` and `unpackTo` are user exposed, I guess we will need to have an API review for these as well?

**Comment by mppf** [comment]
> Instead of typeURL I would like to use something more generic for param name, typeURL is specific to Any I guess.

This part isn't user facing so lots of names should be OK.

> Also, since packFrom and unpackTo are user exposed, I guess we will need to have an API review for these as well?

Yes

**Comment by Aniket21mathur** [comment]
@mppf is there any specific reason we are using `param` over `var`. We need the `packageName` field to compare with the deserialized url and throw an error if they don't match. Since we cannot throw compile time error here, is there any other way `param` will be helpful over `var`? 

Having a string param field with default value in a record makes it's type `recordA("")` instead of `recordA`.  I need the record type in the second form for constructing the `typeUrl`. I am doing something like `(messageObj.type):string` to extract that. Though we can add an additional field like `packageName` to the record to contain the message/record name.

**Comment by mppf** [comment]
@Aniket21mathur - it would be fine to make it `const` instead of `param`. It is true it would make the type generic which isn't necessarily what we want. I was just thinking it could be param, but as you said there is an issue with that.

Part of my thinking here is that instead of constructing the typeURL from `messageObj.type:string` you would simply store it in the record (and have it generated by the protobuf compiler). That seems to avoid potential for issues since the protobuf compiler probably has the easiest time creating an authoritative and correct typeUrl.

**Comment by Aniket21mathur** [comment]
@mppf seems using `const` instead of `param` throws an assignment error ->

```shell
cannot assign to a record of the type test using the default assignment operator because it has 'const' field(s)
test/endToEnd/any/read.chpl:11: Function 'unpa
```

**Comment by mppf** [comment]
@Aniket21mathur - I was thinking about this over the weekend and I think you should just put the generate typeURL (or whatever the field you are choosing) into a function. Like this

``` chapel
record GeneratedMessage {
  proc typeUrl param { return "some.type.url"; }
}
```

Note that `typeUrl` above is called as `myGeneratedMessage.typeUrl` because it is a paren-less method.

**Comment by Aniket21mathur** [comment]
@mppf I have made the above changes. I think the pr is good to go for a final review.

File: src/ProtobufProtocolSupport.chpl [hunk 1]
```diff
@@ -645,6 +645,73 @@ module ProtobufProtocolSupport {
...
+        tmpMem.close();
+
+        this.value = s;
+        this.typeUrl = getTypeUrl(messageObj);
+      }
+
+      proc unpackTo(ref messageObj) throws {
+        var url = getTypeUrl(messageObj);
+        if (url != this.typeUrl) {
+          throw new owned IllegalArgumentError("input message type is not same as global identifier");
```
Comment (mppf) [inline src/ProtobufProtocolSupport.chpl]: reword to maybe "input message type does not match destination message type"

**Review by mppf — APPROVED** [review]

File: src/ProtobufProtocolSupport.chpl (hunk 1 above)
Comment (Aniket21mathur) [reply src/ProtobufProtocolSupport.chpl]: Yeah sure!

**Pull request merged**